query/src/storages/fuse/io/block_reader.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use common_arrow::arrow::datatypes::Schema as ArrowSchema;
+use common_arrow::arrow::io::parquet::read::decompress;
+use common_arrow::arrow::io::parquet::read::page_stream_to_array;
+use common_arrow::arrow::io::parquet::read::read_metadata;
+use common_arrow::arrow::io::parquet::read::read_metadata_async;
+use common_arrow::arrow::io::parquet::read::schema::FileMetaData;
+use common_arrow::parquet::read::get_page_stream;
+use common_dal::DataAccessor;
+use common_datablocks::DataBlock;
+use common_datavalues::prelude::DataColumn;
+use common_datavalues::series::IntoSeries;
+use common_datavalues::DataSchemaRef;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_tracing::tracing;
+use common_tracing::tracing::debug_span;
+use common_tracing::tracing::Instrument;
+use futures::io::BufReader;
+use futures::StreamExt;
+use futures::TryStreamExt;
+
+use crate::storages::fuse::cache::FuseCache;
+
+/// default buffer size of BufferedReader, 1MB
+const DEFAULT_READ_BUFFER_SIZE: u64 = 1024 * 1024;
+
+pub struct BlockReader {
+    data_accessor: Arc<dyn DataAccessor>,
+    path: String,
+
+    block_schema: DataSchemaRef,
+    arrow_table_schema: ArrowSchema,
+    projection: Vec<usize>,
+    row_group: usize,
+    row_groups: usize,
+    metadata: Option<FileMetaData>,
```

Comment:
hi, how about make this field `BlockReader::metadata` typed as  `FileMeataData`, instead of (`Option<_>`)?

Reply:
yes, it seems we have cache metadata inside. thanks for reminding me, I'll fix it.

Reply:
yeah, maybe we cloud pass the metadata to `BlockReader` from `FuseTable::do_read`